### DISCUSSION TITLE
Using official browser extension for static files instead of loading from the server

### DISCUSSION THREAD

**Discussion by hazho** [wagtail/wagtail#9744] [Ideas]
We have thought of loading the static files (from the local machine) instead of (from the server)..
the idea came as an urge of a project that our client is demanding lots of ERP Apps in wagtail admin, and we believe that the we better let the idea to be officially managed by the core team for global editors..!

### the extension does the following:

-   Load proper version of static files from the server of CMS owner (the client's server) but only once, and from that point, as long as the admin user is using the same browser, the loaded static files will be from the local machine.
-   Will be detecting if the server upgraded the wagtail version, hens downloading the new version of files and then serve again from the local machine.
-   Detects is the admin user installed the extension or not yet, therefore shows a notification message with the link for the proper extension link (based on the browser).

### the extension has the following config options:

-  The file version to be beautified or minimized.
-  Turn On/Off auto checking for new versions.
-  ...etc.

### related discussion/issues:
all discussions that related to performance and UX (please tag them in comment)

**Comment by lb-** [comment]
Just curious - why can't browser caching solve the issue here?

I am pretty sure the static assets that come with each Wagtail release have a hash in the URL so the files can be cached heavily.

Maybe something needs to be adjusted in Django or your CDN to ensure the right cache headers are getting added. 

I might be missing something here but I thought a local cache of known resources is what the browser already does.

**Comment by gasman** [comment]
Yep, I think what you're describing is the same as setting a far future [Expires:](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expires) header on static files at the web server level.

**Comment by RealOrangeOne** [comment]
For static file serving like this, I **highly** recommend using [whitenoise](http://whitenoise.evans.io/en/stable/index.html) to achieve this. Once implemented, using the `{% static %}` template tag will result in filenames being used which contain a cache of their content, and their cache headers being set well into the future (as Matthew mentions above).

For an example of whitenoise in action, take a look at the static files being loaded on wagtail.org, most importantly their `cache-control` headers and their filenames. We use whitenoise there and it works great. It's worth noting we also have Cloudflare in front of it, but for this topic that's not relevant - the relevant headers are passed through unmodified.
